File: src/fortran_new/p1_1.f90
!*********************************************
!  ПРОГРАММА 1.1F
!  ТАБЛИЧНЫЙ МЕТОД ОТДЕЛЕНИЯ КОРНЕЙ
! 
!  На примере функции Бесселя первого рода.
!  Входные параметры:
! 	P(1) - порядок функции Бесселя;
! 	P(2) - допустимая погрешность вычисления ряда;
! 	X0, X9 - интервал;
! 	H - шаг.
!*********************************************
Program p1_1

Real(8),dimension(2) :: P
Real(8) :: X, X0, X9, H

PRINT *, 'Начальная точка (X0): '
READ (*,*) X0
PRINT *, 'Конечная точка (X9): '
READ (*,*) X9
PRINT *, 'Шаг: '
Read (*,*) H
PRINT *, 'Порядок функции Бесселя: '
READ (*,*) P(1)
PRINT *, 'Погрешность вычисления ряда: '
READ (*,*) P(2)
X=X0
DO 
  IF (X>X9) THEN
    Exit
  END IF
    
  PRINT *, X, F(X, P)
  X=X+H
END DO

Contains

Real(8) Function F(X, P)

Real(8),dimension(2),intent(in) :: P
Real(8),intent(in) :: X

Real(8) :: R, T
Integer :: N, I

R=X/2 
T=1.0 
N=INT(P(1))
DO I=1, N
  T=T*R/I
END DO
I=1 
R=-R*R
F=T
DO 
  IF (ABS(T)<=P(2)) THEN
    Exit
  END IF
  T=T*R/(I*(I+N)) 
  F=F+T
  I=I+1
END DO

End Function F

End program p1_1

File: src/original/p1_1.pas
{*********************************************
 * ПРОГРАММА 1.1P
 * ТАБЛИЧНЫЙ МЕТОД ОТДЕЛЕНИЯ КОРНЕЙ
 *
 * На примере функции Бесселя первого рода.
 * Входные параметры:
 *	P[1] - порядок функции Бесселя;
 *	P[2] - допустимая погрешность вычисления ряда;
 *	X0, X9 - интервал;
 *	H - шаг.
 *	N - 2 (по количеству ячеек P).
 * На выходе:
 *	таблица значений X и F(X)
 *********************************************}

VAR 
  P: ARRAY[1..9] OF REAL; 
  X, X0, X9, H: REAL;
  N, K: INTEGER;

{ Исследуемая функция }
FUNCTION F(X: REAL): REAL;
VAR 
  R, S, T: REAL; 
  N: INTEGER;
BEGIN 
  R:=X/2; 
  T:=1.0; 
  N:=ROUND(P[1]);
  FOR K:=1 TO N DO 
    T:=T*R/K;
  K:=1; 
  S:=T;
  // Это условие для очень старых советских ЭВМ. 
  // Сейчас оно, наверное, не нужно. Оставить только WHILE.
  IF ABS(R)>1.0E-18 THEN 
  BEGIN 
    R:=-R*R;
    WHILE ABS(T)>P[2] DO 
    BEGIN
      T:=T*R/(K*(K+N)); 
      S:=S+T; 
      K:=K+1;
    END
  END;
  F:=S;
END;

BEGIN
  REPEAT 
    WRITE('X0,X9, H ? ' );
    READLN(X0, X9, H);
    WRITE('СКОЛЬКО ПАРАМЕТРОВ (1..9) ? '); 
    READLN(N);
    FOR K:=1 TO N DO 
    BEGIN
      WRITE('P(', K:2, ') ? '); 
      READ(P[K]);
    END;
    X:=X0;
    WHILE X<=X9 DO 
    BEGIN
      WRITELN(X, ' ', F(X)); 
      X:=X+H;
    END;
    WRITELN;
  UNTIL FALSE;
END.
